ratis-client/src/main/java/org/apache/ratis/client/impl/OrderedAsync.java

```diff
@@ -199,30 +195,26 @@ private void sendRequestWithRetry(PendingOrderedRequest pending) {
       return;
     }
 
-    final RetryPolicy retryPolicy = client.getRetryPolicy();
     sendRequest(pending).exceptionally(e -> {
       if (e instanceof CompletionException) {
         e = JavaUtils.unwrapCompletionException(e);
-        scheduleWithTimeout(pending, request, retryPolicy, e);
+        scheduleWithTimeout(pending, request, e);
         return null;
       }
       f.completeExceptionally(e);
       return null;
     });
   }
 
-  private void scheduleWithTimeout(PendingOrderedRequest pending,
-      RaftClientRequest request, RetryPolicy retryPolicy, Throwable e) {
-    final int attempt = pending.getAttemptCount();
-    final ClientRetryEvent event = new ClientRetryEvent(request, e, pending);
+  private void scheduleWithTimeout(PendingOrderedRequest pending, RaftClientRequest request, Throwable e) {
+    final RetryPolicy retryPolicy = client.getRetryPolicy();
+    final ClientRetryEvent event = pending.newClientRetryEvent(request, e);
     final TimeDuration sleepTime = client.getEffectiveSleepTime(e,
         retryPolicy.handleAttemptFailure(event).getSleepTime());
-    LOG.debug("schedule* attempt #{} with sleep {} and policy {} for {}", attempt, sleepTime, retryPolicy, request);
-    scheduleWithTimeout(pending, sleepTime, getSlidingWindow(request));
-  }
+    LOG.debug("schedule* attempt #{} with sleep {} and policy {} for {}",
+        pending.getAttemptCount(), sleepTime, retryPolicy, request);
```

Comment:
nit: get attempt count from `event` (where it's `final`) instead of `pending`

```suggestion
        event.getAttemptCount(), sleepTime, retryPolicy, request);
```